File: java/Heap.java
/**
 * @author Mike Mirzayanov (mirzayanovmr@gmail.com)
 */
public class Heap {
    private static final int N = 10000000;
    private static final int[] h = new int[N];

    private static void pushDown(int pos, int n) {
        while (2 * pos + 1 < n) {
            int j = 2 * pos + 1;
            if (j + 1 < n && h[j + 1] > h[j])
                j++;
            if (h[pos] >= h[j])
                break;
            int t = h[pos];
            h[pos] = h[j];
            h[j] = t;
            pos = j;
        }
    }

    public static void main(String[] args) {
        long start = System.currentTimeMillis();

        for (int i = 0; i < N; i++) {
            h[i] = i;
        }

        for (int i = N / 2; i >= 0; i--)
            pushDown(i, N);

        int n = N;
        while (n > 1) {
            int t = h[0];
            h[0] = h[n - 1];
            h[n - 1] = t;
            n--;
            pushDown(0, n);
        }

        for (int i = 0; i < N; i++) {
            if (h[i] != i) {
                throw new RuntimeException("h[i] != i");
            }
        }

        System.out.println("Done in " + (System.currentTimeMillis() - start));
    }
}

File: ocaml/heap.ml
(* ocamlopt unix.cmxa -unsafe heap.ml -o heap *)

let _ =
  let n = 10_000_000 in
  let h = Array.init n (fun i -> i) in
  let push_down pos n =
    let pos = ref pos in
      while 2 * !pos + 1 < n do
        let j = 2 * !pos + 1 in
	let j =
          if j + 1 < n && h.(j + 1) > h.(j)
          then j + 1
	  else j
	in
          if h.(!pos) >= h.(j)
	  then pos := n
          else (
	    let tmp = h.(!pos) in
	      h.(!pos) <- h.(j);
	      h.(j) <- tmp;
              pos := j;
	  )
      done
  in
  let start = Unix.gettimeofday () in
    for i = n / 2 downto 0 do
      push_down i n
    done;
    for i = n - 1 downto 1 do
       let tmp = h.(0) in
	 h.(0) <- h.(i);
	 h.(i) <- tmp;
	 push_down 0 i;
    done;
    for i = 0 to n - 1 do
      assert (h.(i) = i)
    done;
    Printf.printf "Done in %f\n" (Unix.gettimeofday () -. start)


